utility/lambdas/ecs-asg-sizer/lambda_function.py

```diff
@@ -210,22 +224,62 @@ def get_usages(cluster_name):
     instance_arns = instance_list['containerInstanceArns']
     details = ecs_client.describe_container_instances(
         cluster=cluster_name, containerInstances=instance_arns)
-    return list(map(get_usage, details['containerInstances']))
+    usages = list(map(get_usage, details['containerInstances']))
+
+    # mark elapsed time agents have been disconnected
+    cluster_arn = get_cluster_arn(cluster_name)
+    disconnects = get_disconnects(cluster_arn)
+    now = int(datetime.datetime.now().timestamp())
+    for usage in usages:
+        if usage['disconnected'] and disconnects.get(usage['id']):
+            usage['disconnectseconds'] = now - disconnects.get(usage['id'])
+    set_disconnects(cluster_arn, disconnects, usages, now)
+
+    return usages
 
 
 def get_usage(details):
     rem = details['remainingResources']
     reg = details['registeredResources']
     return {
         'id': details['ec2InstanceId'],
-        'connected': details['agentConnected'],
+        'disconnected': details['agentConnected'] == False,
+        'disconnectseconds': 0,
         'cpu': next(r['integerValue'] for r in rem if r['name'] == 'CPU'),
         'mem': next(r['integerValue'] for r in rem if r['name'] == 'MEMORY'),
         'tcpu': next(r['integerValue'] for r in reg if r['name'] == 'CPU'),
         'tmem': next(r['integerValue'] for r in reg if r['name'] == 'MEMORY'),
     }
 
 
+def get_cluster_arn(cluster_name):
+    clusters = ecs_client.describe_clusters(clusters=[cluster_name])
+    return clusters['clusters'][0]['clusterArn']
+
+
+def get_disconnects(cluster_arn):
+    tags = ecs_client.list_tags_for_resource(resourceArn=cluster_arn)
+    disconnects = {}
+    for tag in tags['tags']:
+        if tag['key'] == DISCONNECTS_TAG:
+            for value in tag['value'].split(DISCONNECTS_DELIMITER):
+                parts = value.split(DISCONNECTS_SEPARATOR)
+                if len(parts) == 2:
+                    disconnects[parts[0]] = int(parts[1])
+    return disconnects
+
+
+def set_disconnects(cluster_arn, previous_disconnects, usages, now):
+    disconnects = []
+    for usage in usages:
+        if usage['disconnected']:
```

Comment:
Does this need the other side?  In the case where an instance comes back and then flickers off again -- having a stale `disconnectseconds` from the earlier flicker.  Maybe that never happens!

Reply:
Yep - that should be covered.  The cases here are:

1. Instance disconnects for the 1st time - add to the `disconnects` hash.
2. Instance disconnects for a subsequent time - keep old timestamp in `disconnects` hash.
3. Instance _was_ disconnected, but this time around, it's connected.  Remove from the `disconnects` hash.  (Which happens implicitly because I'm rebuilding the list here).

I'm going to take another shot at organizing/commenting this method ... see if I can't make that clearer.